src/software/backend/simulation/simulator.cpp

```diff
@@ -1,63 +1,97 @@
- * because it doesn't know the size of the object. Therefore we need to create our own
- * "Deleter" class we can provide to the smart pointers to handle that instead.
- *
- * See https://en.cppreference.com/w/cpp/memory/unique_ptr/unique_ptr for more info and
- * examples
- */
-struct FirmwareWorldDeleter
-{
-    void operator()(FirmwareWorld_t* firmware_world) const
-    {
-        app_firmware_world_destroy(firmware_world);
-    };
-};
-
-
 Simulator::Simulator(const World& world) : physics_world(world)
 {
     for (auto physics_robot : physics_world.getFriendlyPhysicsRobots())
     {
         auto simulator_robot = std::make_shared<SimulatorRobot>(physics_robot);
-        friendly_simulator_robots.emplace_back(simulator_robot);
+        std::unique_ptr<PrimitiveManager, PrimitiveManagerDeleter> primitive_manager(
+            app_primitive_manager_create(), PrimitiveManagerDeleter());
+        this->simulator_robots.emplace(simulator_robot, std::move(primitive_manager));
     }
     simulator_ball = std::make_shared<SimulatorBall>(physics_world.getPhysicsBall());
-}
 
-void Simulator::stepSimulation(const Duration& time_step)
-{
     auto firmware_robot = SimulatorRobotSingleton::createFirmwareRobot();
     auto firmware_ball  = SimulatorBallSingleton::createFirmwareBall();
+    // Make sure to release ownership from the local variables so the firmware_world
+    // can take ownership
     FirmwareWorld_t* firmware_world_raw =
-        app_firmware_world_create(firmware_robot.get(), firmware_ball.get());
-    auto firmware_world = std::unique_ptr<FirmwareWorld_t, FirmwareWorldDeleter>(
+        app_firmware_world_create(firmware_robot.release(), firmware_ball.release());
+    firmware_world = std::unique_ptr<FirmwareWorld_t, FirmwareWorldDeleter>(
         firmware_world_raw, FirmwareWorldDeleter());
+}
+
+void Simulator::stepSimulation(const Duration& time_step)
+{
     SimulatorBallSingleton::setSimulatorBall(simulator_ball);
-    for (const auto& simulator_robot : friendly_simulator_robots)
+    for (auto iter = simulator_robots.begin(); iter != simulator_robots.end(); iter++)
     {
+        auto simulator_robot    = iter->first;
+        auto& primitive_manager = iter->second;
         SimulatorRobotSingleton::setSimulatorRobot(simulator_robot);
-        // TODO: Simulate primitives
-        // app_main(firmware_world.get(), PRIMITIVE STUFF);
+        app_primitive_manager_runCurrentPrimitive(primitive_manager.get(),
+                                                  firmware_world.get());
     }
 
     physics_world.stepSimulation(time_step);
 }
 
 void Simulator::setPrimitives(ConstPrimitiveVectorPtr primitives)
 {
-    // TODO: Set primitives
-    // https://github.com/UBC-Thunderbots/Software/issues/768
+    if (primitives)
+    {
+        auto& primitive_vector = *primitives;
+        for (auto iter = simulator_robots.begin(); iter != simulator_robots.end(); iter++)
```

Comment:
Can we change to `for (auto& pair : simulator_robots)`?

Reply:
done